lib/engine/game/base.rb

```diff
@@ -192,18 +188,77 @@ def active_players
         @round.active_entities.map(&:owner)
       end
 
+      # Initialize actions respecting the undo state
+      def initialize_actions(actions)
+        active_undos = []
+        filtered_actions = Array.new(actions.size)
+
+        actions.each.with_index do |action, index|
+          case action['type']
+          when 'undo'
+            i = filtered_actions.rindex { |a| !a.nil? && a['type'] != 'message' }
+            active_undos << [filtered_actions[i], i]
+            filtered_actions[i] = nil
+          when 'redo'
+            a, i = active_undos.pop
+            filtered_actions[i] = a
+          when 'message'
+            # Messages do not get undoed.
+            # warning adding more types of action here will break existing game
+            filtered_actions[index] = action
+          else
+            active_undos = []
+            filtered_actions[index] = action
+          end
+        end
+
+        @undo_possible = false
+        # replay all actions with a copy
+        filtered_actions.each.with_index do |action, index|
+          if action.nil?
+            # Restore the original action to the list to ensure action ids remain consistent but don't apply them
+            @actions << actions[index]
+          else
+            action = action.copy(self) if action.is_a?(Action::Base)
+            process_action(action)
+          end
+        end
+        @redo_possible = !active_undos.empty?
+      end
+
+      def can_undo?
```

Comment:
i'd remove these two helpers and just make redo_possible and can_undo as readers